README.md
# sls-utils

[![Maintainability](https://api.codeclimate.com/v1/badges/7f7ce6f918e9fba52b69/maintainability)](https://codeclimate.com/github/powerkernel/sls-utils/maintainability)
[![Test Coverage](https://api.codeclimate.com/v1/badges/7f7ce6f918e9fba52b69/test_coverage)](https://codeclimate.com/github/powerkernel/sls-utils/test_coverage)
[![Build Status](https://travis-ci.com/powerkernel/sls-utils.svg?branch=master)](https://travis-ci.com/powerkernel/sls-utils)
[![Dependency Status](https://img.shields.io/david/powerkernel/sls-utils.svg?style=flat-square)](https://david-dm.org/powerkernel/sls-utils)
[![Download Status](https://img.shields.io/npm/dm/@powerkernel/sls-utils.svg?style=flat-square)](https://www.npmjs.com/package/@powerkernel/sls-utils)

> Power Kernel utils for serverless

## Installation

```
npm install @powerkernel/sls-utils --save
```

## Documentation

Common utils to use with serverless framework

## Available utils

The following utils are svailable:
`response`

Auto add headers

```js
headers: {
    'Access-Control-Allow-Origin': '*',
    'Access-Control-Allow-Credentials': true,
}
```

Success response

```js
response.success = ({ headers = {}, code = 200, data = {} });
```

Failure response

```js
response.failure = ({ headers = {}, code = 500, error = {} });
```

## Complete Example

Here's an example of a serverless function that uses `response` to output the response.

```js
import { response } from "@powerkernel/sls-utils";

import uuid from "uuid";
import * as dynamoDbLib from "./libs/dynamodb-lib";

export async function main(event, context) {
  const data = JSON.parse(event.body);
  const params = {
    TableName: process.env.tableName,
    Item: {
      userId: event.requestContext.identity.cognitoIdentityId,
      noteId: uuid.v1(),
      content: data.content,
      attachment: data.attachment,
      createdAt: Date.now()
    }
  };

  try {
    await dynamoDbLib.call("put", params);
    return response.success({ data: params.Item});
  } catch (e) {
    return response.failure({ error: { message: "Something went wrong." }});
  }
```

## License

Copyright (c) 2020 Power Kernel

Licensed under the [MIT license](LICENSE).


CHANGELOG.md
### Changelog

All notable changes to this project will be documented in this file. Dates are displayed in UTC.

Generated by [`auto-changelog`](https://github.com/CookPete/auto-changelog).

#### [v1.1.1](https://github.com/powerkernel/sls-utils/compare/v1.1.0...v1.1.1)

> 20 March 2020

- update README + github action [`cf87b4a`](https://github.com/powerkernel/sls-utils/commit/cf87b4a56df4398c3c99aa16c736c2e333451cc2)

#### [v1.1.0](https://github.com/powerkernel/sls-utils/compare/1.0.0...v1.1.0)

> 20 March 2020

- convert to TypeScript [`92e9fd8`](https://github.com/powerkernel/sls-utils/commit/92e9fd8da3ad7c2cea501e08f47b8a4aa5918ee9)
- Create npmpublish.yml [`9f73c89`](https://github.com/powerkernel/sls-utils/commit/9f73c89c36dd4e61e40cd1846c1e89cc0d553e78)
- update username [`41a1353`](https://github.com/powerkernel/sls-utils/commit/41a1353f4341c8fccb8a22101bed519530374ad0)

#### 1.0.0

> 14 March 2020

- first commit [`8a8ee6c`](https://github.com/powerkernel/sls-utils/commit/8a8ee6ccbf37716ff5bc2e767b5f4d1240f6455c)
- releasing version [`a56b218`](https://github.com/powerkernel/sls-utils/commit/a56b21803b40921bb67f0d65d191fe945db35b6c)


src/response.test.js
/*
 * @Author: Harry Tang - harry@powerkernel.com
 * @Date: 2020-03-14 14:22:40
 * @Last Modified by: Harry Tang - harry@powerkernel.com
 * @Last Modified time: 2020-03-14 16:46:46
 */

const { response } = require('./index');

describe('Response', () => {
  it('returns a success response', () => {
    const res = response.success({
      data: 'This is a success message',
    });

    expect(res).toEqual({
      statusCode: 200,
      headers: {
        'Access-Control-Allow-Origin': '*',
        'Access-Control-Allow-Credentials': true,
      },
      body: '{"success":true,"data":"This is a success message"}',
    });
  });

  it('returns a success response with custom headers', () => {
    const res = response.success({
      headers: { 'x-power-by': 'Power Kernel' },
      data: 'Custom header added',
    });

    expect(res).toEqual({
      statusCode: 200,
      headers: {
        'x-power-by': 'Power Kernel',
        'Access-Control-Allow-Origin': '*',
        'Access-Control-Allow-Credentials': true,
      },
      body: '{"success":true,"data":"Custom header added"}',
    });
  });

  it('returns a failure response', () => {
    const res = response.failure({
      code: 404,
      error: 'This is a failure message',
    });
    expect(res).toEqual({
      statusCode: 404,
      headers: {
        'Access-Control-Allow-Origin': '*',
        'Access-Control-Allow-Credentials': true,
      },
      body: '{"success":false,"error":"This is a failure message"}',
    });
  });
});


test/response.test.ts
/*
 * @Author: Harry Tang - harry@powerkernel.com
 * @Date: 2020-03-14 14:22:40
 * @Last Modified by: Harry Tang - harry@powerkernel.com
 * @Last Modified time: 2020-03-14 16:46:46
 */

import { response } from "../src/index";

describe("Response", () => {
  it("returns a success response", () => {
    const res = response.success({
      data: {
        message: "This is a success message"
      }
    });

    expect(res).toEqual({
      statusCode: 200,
      headers: {
        "Access-Control-Allow-Origin": "*",
        "Access-Control-Allow-Credentials": true
      },
      body: "{\"success\":true,\"data\":{\"message\":\"This is a success message\"}}"
    });
  });

  it("returns a success response with custom headers", () => {
    const res = response.success({
      headers: { "x-power-by": "Power Kernel" },
      data: {
        message: "Custom header added"
      }
    });

    expect(res).toEqual({
      statusCode: 200,
      headers: {
        "x-power-by": "Power Kernel",
        "Access-Control-Allow-Origin": "*",
        "Access-Control-Allow-Credentials": true
      },
      body: "{\"success\":true,\"data\":{\"message\":\"Custom header added\"}}"
    });
  });

  it("returns a failure response", () => {
    const res = response.failure({
      code: 404,
      error: {
        message: "This is a failure message"
      }
    });
    expect(res).toEqual({
      statusCode: 404,
      headers: {
        "Access-Control-Allow-Origin": "*",
        "Access-Control-Allow-Credentials": true
      },
      body: "{\"success\":false,\"error\":{\"message\":\"This is a failure message\"}}"
    });
  });

  it("returns a failure response with custom header", () => {
    const res = response.failure({
      headers: { "x-power-by": "Power Kernel" },
      error: {
        message: "This is a failure message"
      }
    });
    expect(res).toEqual({
      statusCode: 500,
      headers: {
        "x-power-by": "Power Kernel",
        "Access-Control-Allow-Origin": "*",
        "Access-Control-Allow-Credentials": true
      },
      body: "{\"success\":false,\"error\":{\"message\":\"This is a failure message\"}}"
    });
  });
});


src/response.js
/*
 * @Author: Harry Tang - harry@powerkernel.com
 * @Date: 2020-03-14 14:08:50
 * @Last Modified by: Harry Tang - harry@powerkernel.com
 * @Last Modified time: 2020-03-14 16:36:24
 */

const response = {};

const buildResponse = ({
  headers,
  code,
  success,
  data = {},
  error = {},
}) => ({
  statusCode: code,
  headers: {
    ...headers,
    'Access-Control-Allow-Origin': '*',
    'Access-Control-Allow-Credentials': true,
  },
  body: success
    ? JSON.stringify({
      success,
      data,
    })
    : JSON.stringify({
      success,
      error,
    }),
});

response.success = ({ headers = {}, code = 200, data = {} }) => buildResponse({
  headers, code, success: true, data,
});

response.failure = ({ headers = {}, code = 500, error = {} }) => buildResponse({
  headers, code, success: false, error,
});

module.exports = response;


src/response.ts
/*
 * @Author: Harry Tang - harry@powerkernel.com
 * @Date: 2020-03-14 14:08:50
 * @Last Modified by: Harry Tang - harry@powerkernel.com
 * @Last Modified time: 2020-03-14 16:36:24
 */

interface HeaderData {
  [key: string]: string | true | false;
}

interface ResData {
  // eslint-disable-next-line
  [key: string]: any;
}

interface SuccessInput {
  headers?: HeaderData;
  code?: number;

  data: ResData;
}

interface FailureInput {
  headers?: HeaderData;
  code?: number;
  error: ResData;
}

interface Response {
  statusCode: number;
  headers: HeaderData;
  body: string;
}

const buildHeaders = (headers: HeaderData): HeaderData => {
  return {
    ...headers,
    "Access-Control-Allow-Origin": "*",
    "Access-Control-Allow-Credentials": true
  };
};

const success = ({
  headers = {},
  code = 200,
  data
}: SuccessInput): Response => {
  return {
    statusCode: code,
    headers: buildHeaders(headers),
    body: JSON.stringify({
      success: true,
      data
    })
  };
};

const failure = ({
  headers = {},
  code = 500,
  error
}: FailureInput): Response => {
  return {
    statusCode: code,
    headers: buildHeaders(headers),
    body: JSON.stringify({
      success: false,
      error
    })
  };
};

const response = {
  success,
  failure
};
export default response;
